File: sdk/elixir/lib/direnv_config/store.ex
defmodule DirenvConfig.Store do
  @spec state_dir() :: String.t()
  def state_dir do
    case System.get_env("XDG_STATE_HOME") do
      nil -> Path.join([System.user_home!(), ".local", "state", "direnv-config"])
      xdg -> Path.join(xdg, "direnv-config")
    end
  end

  @spec path_to_hash(String.t()) :: String.t()
  def path_to_hash(dir) do
    stripped = String.trim_leading(dir, "/")
    name = String.replace(stripped, "/", "-")

    if byte_size(name) <= 200 do
      name
    else
      hash =
        :crypto.hash(:sha256, dir)
        |> Base.encode16(case: :lower)
        |> binary_part(0, 8)

      binary_part(name, 0, 200) <> "-" <> hash
    end
  end

  @spec store_path(String.t()) :: String.t()
  def store_path(dir) do
    Path.join(state_dir(), path_to_hash(dir))
  end

  @spec find_current_store(String.t() | nil) :: {:ok, String.t()} | {:error, :not_found}
  def find_current_store(start_dir \\ nil) do
    dir = start_dir || File.cwd!()
    walk_up(dir)
  end

  defp walk_up("/"), do: check_or_stop("/")

  defp walk_up(dir) do
    sp = store_path(dir)

    if File.dir?(sp) do
      {:ok, sp}
    else
      parent = Path.dirname(dir)

      if parent == dir do
        {:error, :not_found}
      else
        walk_up(parent)
      end
    end
  end

  defp check_or_stop(dir) do
    sp = store_path(dir)
    if File.dir?(sp), do: {:ok, sp}, else: {:error, :not_found}
  end
end

File: sdk/python/src/direnv_config/store.py
from __future__ import annotations

import hashlib
import os
from pathlib import Path


class StoreNotFoundError(Exception):
    pass


def state_dir() -> Path:
    xdg = os.environ.get("XDG_STATE_HOME")
    if xdg:
        return Path(xdg) / "direnv-config"
    home = Path.home()
    return home / ".local" / "state" / "direnv-config"


def path_to_hash(directory: str) -> str:
    stripped = directory.lstrip("/")
    name = stripped.replace("/", "-")

    if len(name) <= 200:
        return name

    digest = hashlib.sha256(directory.encode()).hexdigest()
    return f"{name[:200]}-{digest[:8]}"


def store_path(directory: str) -> Path:
    return state_dir() / path_to_hash(directory)


def find_current_store(start_dir: str | None = None) -> Path:
    current = Path(start_dir) if start_dir else Path.cwd()
    current = current.resolve()

    while True:
        sp = store_path(str(current))
        if sp.exists():
            return sp
        parent = current.parent
        if parent == current:
            break
        current = parent

    raise StoreNotFoundError(
        f"no store found for {start_dir or Path.cwd()} "
        f"(searched all parent directories). Run `dc init` first."
    )
